h/services/annotation_metadata.py

```diff
@@ -0,0 +1,48 @@
+import os
+
+from sqlalchemy.dialects.postgresql import insert
+from sqlalchemy.orm import Session
+
+from h.models import Annotation, AnnotationMetadata, User
+from h.security import decrypt_jwe_dict
+
+
+class AnnotationMetadataService:
+    def __init__(self, db: Session):
+        self._db = db
+        self._secret = os.environ.get("JWE_SECRET_LMS")
+
+    def set_annotation_metadata_from_jwe(self, annotation: Annotation, jwe: str):
+        """
+        Set the decrypted `jwe` dict as the metadata for `annotation`.
+
+        It will create a new AnnotationMetadata row or update the existing one.
+        """
+        if not annotation.pk and not annotation.user_id:
+            # Temporarily add these to the annotation until
+            # h.task.annotations.fill_pk_and_user_id completes
+            annotation.pk = Annotation.pk_sequence.next_value()
+            annotation.user_id = (
+                self._db.query(User).filter_by(userid=annotation.userid).one().id
+            )
+            self._db.flush()
+
+        annotation_metadata = decrypt_jwe_dict(self._secret, jwe)
+
+        self._set_annotation_metadata(annotation, annotation_metadata)
+        return annotation_metadata
+
+    def _set_annotation_metadata(self, annotation, metadata: dict):
```

Comment:
I reckon `insert on conflict` is the right approach here.

It reads a bit worse than the pure ORM version but it should safer and more performant.

I reckon the mistake is trying to ORM-ize this. This is a SQL statement, we use the SQLA core API and that's it. No need to create a big wrapper around it.